services/src/main/java/org/keycloak/authentication/requiredactions/DeleteAccount.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.authentication.requiredactions;
+
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.authentication.AuthenticationProcessor;
+import org.keycloak.authentication.InitiatedActionSupport;
+import org.keycloak.authentication.RequiredActionContext;
+import org.keycloak.authentication.RequiredActionFactory;
+import org.keycloak.authentication.RequiredActionProvider;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.AccountRoles;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserManager;
+import org.keycloak.models.UserModel;
+import org.keycloak.services.ForbiddenException;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.messages.Messages;
+
+public class DeleteAccount implements RequiredActionProvider, RequiredActionFactory {
```

Comment:
Only AIA's can do a cancel, a required action is something an admin requires a user to do. Sure, it makes little sense for an admin to add the delete action to the user, but I'd say it's out of the scope for this PR to filter what actions can be used where.

So I would update to not show the delete button as then it's consistent with other actions.

Reply:
> So I would update to not show the delete button as then it's consistent with other actions.

You mean hide the Cancel button unless it's AIA?

Reply:
Good catch. If I remember well, we discussed this on the mailing list. Maybe it would make sens in the future to have AIA only actions. As a workaround now, I just added a check for `context.getAuthenticationSession().getClientNote(Constants.KC_ACTION)`, if is set to `delete_action`, then I display the cancel button, otherwise just the confirm button alone.

Reply:
@zak905 Thanks for the fix! I think the last thing missing here is a test for the required action (also with the corner case when the user is missing the delete-account role) as I mentioned above.